Bluetooth: Mesh: Add support for MshPRTd1.1r18, MshMBT_d1.0r04_PRr00, MshDFU_d1.0r04_PRr00 specs

PavelVPV opened the pull request:
This PR adds support for MshPRTd1.1r18, MshMBT_d1.0r04_PRr00, MshDFU_d1.0r04_PRr00 draft specifications. The details are present below.

### New features

From the Mesh Protocol d1.1r18 specification (https://www.bluetooth.com/specifications/specs/mesh-protocol-1-1/):
- Enhanced Provisioning Authentication support
- Mesh Remote Provisioning support including:
  - Remote Provisioning Server and Client models
  - Composition Data Page 128 and Models Metadata Page 128 support
- Large Composition Data support including:
  - Large Composition Data Server and Client models
  - Models Metadata Page 0 support
- New Transport Segmentation and Reassembly (SAR) implementation including:
  - SAR Configuration Server and Client models
- Mesh Private Beacons support including:
  - Mesh Private Beacon Server and Client models
- Opcodes Aggregator support including:
  - Opcodes Aggregator Server and Client models
- Proxy Solicitation support including:
  - Solicitation PDU RPL Configuration Server and Client models
  - On-Demand Private Proxy Server and Client models
- Composition Data Page 1 support
- Other Mesh Profile Enhancements

From the Mesh Binary Large Object Transfer Model d1.0r04_PRr00 specification (https://www.bluetooth.com/specifications/specs/mesh-binary-large-object-transfer-model-2/):
- BLOB Transfer Server & Client models

From the Mesh Device Firmware Update Model d1.0r04_PRr00 specification (https://www.bluetooth.com/specifications/specs/mesh-device-firmware-update-model-2/):
- Firmware Update Server and Client models
- Firmware Distribution Server model

#### Compatibility with the Mesh Profile 1.0.1 implementation

The new features (including new transport SAR implementation) are enabled by the `CONFIG_BT_MESH_V1d1` Kconfig option thus not affecting qualification of the Mesh Profile 1.0.1 implementation.

#### Shell support

This PR adds shell support for the all features listed above.
The new models support is added to the `mesh_shell` application.

#### Documentation

The documentation is present for all features listed above except:
- SAR Configuration models
- Large Composition Data models
- Remote Provisioning

### BabbleSim tests

#### Current tests

All currently implemented BabbleSim tests run for both the Mesh Profile 1.0.1 and the Mesh Protocol d1.1r18 implementations.

#### New tests

This PR implements BabbleSim tests for the following features:
- Mesh Remote Provisioning
- Large Composition Data
- Opcodes Aggregator
- Transport SAR Configuration
- BLOB Transfer Server & Client models
- Firmware Update Server and Client models
- Firmware Distribution Server model

trond-snekvik committed:
Bluetooth: Mesh: Remote provisioning

Adds remote provisioning models rpr_srv and rpr_cli, and exposes them as
a bearer for the provisioning process.

The remote provisioning models provide the following features:
- Remote provisioning of new devices, using a remote provisioning server
  as a proxy.
- Reprovisioning of existing devices, generating a new device key and
  changing addresses and composition data.

Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Co-authored-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: Add composition data page 128

Composition data page 128 represents the new composition data after a
composition refresh procedure. The implementation stores the old
composition data (before applying a DFU or similar), and uses that as
page 0 if present. As the device has already rebooted by the time page
128 becomes active, its active composition data is page 128, while the
backed up data is the "old" data, which should be shown as page 0.

Co-authored-by: Ingar Kulbrandstad <ingar.kulbrandstad@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: Shell support for remote provisioning

Adds commands in the mesh shell for remote provisioning client
operation, including scanning, provisioning and reprovisioning.

Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: Private Beacons

Adds support for private beacon sending and receiving.

Co-authored-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>
Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: BLOB models

Adds support for the BLOB Models from Mesh Model specification v1.1.

Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Co-authored-by: Anders Storrø <anders.storro@nordicsemi.no>
Co-authored-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>
Co-authored-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>
Co-authored-by: Ludvig Samuelsen Jordet <ludvig.jordet@nordicsemi.no>
Co-authored-by: Mia Koen <mia.koen@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: DFU models

Adds DFU models from the Mesh Model specification v1.1.

Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Co-authored-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>
Co-authored-by: Anders Storrø <anders.storro@nordicsemi.no>
Co-authored-by: Mia Koen <mia.koen@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: DFD Server model

Adds support for the Device Firmware Distribution Server model from Mesh
Model specification v1.1.

Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Co-authored-by: Ludvig Samuelsen Jordet <ludvig.jordet@nordicsemi.no>
Co-authored-by: Anders Storrø <anders.storro@nordicsemi.no>
Co-authored-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>
Co-authored-by: Mia Koen <mia.koen@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

Commit:
Bluetooth: Mesh: Fix Heartbeat Pub/Sub Count/Period Log calculation

Mesh Profile Specification Errata 11737 specifies that
Period/Count Log value 0x11 is not defined Log field
transformation in Table 4.1.

Errata 11737 adds value 0x11 to Log field transformation
table with 2-octet value equal to 0x10000 which does not fit into
two octets. This patch changes period calculation to 32-bit precision.
There is also a special case for publication count. If the
log value is equal to 0x11 then the publication count value
should be 0xfffe. When sending Hearbeat Publication Status
message we should return the same value as the one received
in Publication Set so we have to add a case in hb_pub_count_log().

Co-authored-by: Alperen Sener <alperen.sener@nordicsemi.no>
Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

Commit:
Bluetooth: Mesh: Set minimum publish period

Mesh Profile Errata 11392

Section 4.4.3.2.1 Current Fault state:

When the value of a Health FastPublish Period Divisor state is non-zero,
and a Current Fault Fault Array (see Section 4.2.15.1) for at least one
Company ID contains records, an unsolicited Health Current Status
message set to the value of that Company ID and the FaultArray field
containing a sequence of faults representing a sequence of faults in the
Current Fault Fault Array (see Section 4.2.15.1) shall be published
every number of seconds as defined by the value of the Publish Period
divided by the value represented by the Health Fast Period Divisor
state, or every 100 ms (the minimum Publish Period value whichever
is greater.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
Bluetooth: Mesh: Change IV Update duration requirement for new nodes

Mesh Profile specification errata 11627

Section 3.10.5:
When a node is added to a network, the node is given an IV Index.
If the node is added to a network when the network is in Normal
operation, then it shall operate in Normal operation for at least
96 hours. If a node is added to a network while the network is in
the IV Update in Progress state, then the node shall be given the
new IV Index value and operate in IV Update in Progress operation
without the restriction of being in this state for at least 96 hours.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
Bluetooth: Mesh: Check for duplicates in friend queue

Mesh Profile Specification Errata 11302

Section 3.5.5.

When a Friend node receives a message that is destined for a Low Power
node (i.e., the destination of the message is a unicast address of an
element of the Low Power node or in the Friend Subscription List), and
the TTL field has a value of 2 or greater, then the message shall be
processed as follows: If the Friend Queue already contains a message
with the same SEQ and SRC fields as in the received message, or if the
SRC field of the received message is a unicast address of an element of
the Low Power node, then the message shall not be stored in the Friend
Queue. Otherwise, the TTL field value shall be decremented by 1, and
the message shall be stored in the Friend Queue.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
Bluetooth: Mesh: Do not relay message rejected by Transport layer

Mesh Profile Specification Errata 11341

3.6.4.3 Message error procedure
When the Upper Transport layer receives a message that is not
understood, then the message shall be ignored.
A message that is not understood includes messages that met one or more
conditions listed below:
• The Transport Control message opcode is unknown by the receiving node.
• The Transport Control message size for the Transport Control opcode
is incorrect.
• The Transport Control message parameters contain values that are
Prohibited

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
Bluetooth: Mesh: Add Model Receive testing callback

Add a callback similar to net_recv that notifies the application when a
message is received in access layer. This is is useful for testing
purposes (especially automated testing).

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
Bluetooth: Mesh: Refactor bt_mesh_model_recv

Pass only msg context as it is the only thing used from net_rx.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
Bluetooth: Mesh: Add a flag for DevKey-only based models

This prevents a Configuration Client from binding an app key to a model
that only supports DevKey-based security.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

Commit:
Bluetooth: Mesh: Add support for Large Composition Data models

- Adds Large Composition Data Server and Client definitions
- Adds Client API definition
- Adds Health Server Metadata definition
- Refactors Composition Data processing (as a Server) to use common
 methods with Large Comp Data Server

Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Co-authored-by: Stine Akredalen <stine.akredalen@nordicsemi.no>
Co-authored-by: Omkar Kulkarni <omkar.kulkarni@nordicsemi.no>
Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

Commit:
shell: Add Large Composition Data Server

- Adds Large Comp Data Server and Client
- Defines some dummy metadata for Health Server

Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

Commit:
Bluetooth: Mesh: Add support for Opcodes Aggregator models

Includes Opcodes Aggregator Server and Client. To use Opcodes Aggregator
client features with some client model, that client should support async
API.

Co-authored-by: Ludvig Samuelsen Jordet <ludvig.jordet@nordicsemi.no>
Co-authored-by: Mia Koen <mia.koen@nordicsemi.no>
Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

Commit:
Bluetooth: Mesh: Add Transport SAR Configuration models to shell

Add Transport SAR Configuration models to shell.

Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

Commit:
Bluetooth: Mesh: Add Opcodes Aggregator models to shell

- Add entry for client and server in composition data.
- Add Client commands to aggregate and send messages.

MESH/SR/AGG/BV-02-C requires more key slots.

Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

alxelax committed:
Bluetooth: Mesh: enhanced provisioning authentication

Implementation of the enhanced provisioning authentication (EPA)
that is based on SHA256 and HMAC security algorithms.
OOB authentication mandates EPA usage.

Co-authored-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>
Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

Andrewpini committed:
Tests: Bluetooth: Mesh: BLOB cli mod ret caps test

Add bsim test for retrieve capabilities of the BLOB client model.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Tests: Bluetooth: Mesh: BLOB cli mod broadcast API

Add bsim test for broadcast API of the BLOB client model.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Tests: Bluetooth: Mesh: BLOB cli ret caps suspend

Add bsim test for suspending retrieve capabilities procedure of the BLOB
client model.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Tests: Bluetooth: Mesh: BLOB cli broadcast unicast

Add bsim test using unicast addr for broadcast API of the BLOB client
model.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Tests: Bluetooth: Mesh: BLOB cli push & pull trans resume

Add bsim test for resuming push & pull mode transfer on the BLOB client
after BLOB server has been suspended during initial attempt.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

PavelVPV committed:
Bluetooth: Mesh: Add shell commands for Private Beacon Client

Add shell commands for Private Beacon Client.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

Andrewpini committed:
Tests: Bluetooth: Mesh: BLOB cli transfer persist

Tests that BLOB Client continues the Push BLOB Transfer procedure
as long as there is at least one active target.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Tests: Bluetooth: Mesh: BLOB cli no response

- Tests that BLOB Client suspends if no target responds to Block get.
- Tests that BLOB Client transfer ends if no target responds to Xfer
  get.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Commit:
Bluetooth: Mesh: Add support for Models Metadata Page 128

This patch adds for handling Models Metadata Get Page 128.

Signed-off-by: Michal Narajowski <michal.narajowski@codecoup.pl>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
tests: Bluetooth: Mesh: Add tests for BLOB Transfer in Pull mode

This commit adds several BabbleSim tests for BLOB transfer:
- A sanity test checking that BLOB Transfer completes successfully in
both, Push and Pull modes that runs transfer for 4 targets
- blob_cli_trans_resume test for Pull mode
- Test checking that BLOB client continues BLOB transfer if one of
targets times out while sending chunks.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

KKopyscinski committed:
Bluetooth: Mesh: Add Composition Data Page 1

Composition Data Page 1 contains information about relations between
models. Implementation saves relations into array of tuples
{ptr to base, ptr to extension, relation type} on model initialization
and uses this information to construct Composition Data Page 1.

Appropriate structures and methods were added to API, allowing user to
parse all received data into Composition Data Page 1 formats:
Composition Data Page 1 Element, Model Item, Model Extension Item.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

KKopyscinski committed:
Bluetooth: Mesh: Add Composition Data Page 1 to shell

Add parsing of Composition Data Page 1 and enable it's support.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

PavelVPV committed:
tests: Bluetooth: Mesh: Test reprovisioning through RPR models

Test that Remote Provisioning models can re-provision a remote node
multiple times.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
tests: bluetooth: mesh: Implement robustness test for NPPI procedures

Test Dev Key Refresh, Node Composition Refresh and Node Address Refresh
procedures for robutness.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

KKopyscinski committed:
Bluetooth: Mesh: add implementation for Proxy Solicitation

This is initial implementation of Proxy solicitation procedure.
This includes:
- support for sending and receiving Solicitation PDUs
- On-Demand Private Proxy functionality (Server and Client) controlling
  behaviour of node after receiving Solicitation PDU
- Solicitation PDU RPL Configuration (Server and Client), which manages
  Replay Protection List for Solicitation PDUs.

Proxy Solicitation allows to enable advertising of Proxy service on node
by sending Solicitation PDUs. These PDUs are not part of Mesh messages;
instead, these are non-connectable, undirected advertising PDUs with
their own format, containing Proxy Solicitation UUID.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

akredalen committed:
Bluetooth: mesh: added OPcode aggregator Bsim test

The test checks the Opcode aggregator (SIG) model basic behaviour.

Signed-off-by: Stine Akredalen <stine.akredalen@nordicsemi.no>

PavelVPV committed:
tests: Bluetooth: Mesh: Test Node Composition refresh with settings

Test Node Composition Refresh procedure with persistence settings.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

akredalen committed:
tests: Bluetooth: Mesh: added SAR stress test

Test that a maximum length SDU can be processed with SAR
with "stressed" transmitter and receiver configurations.

Signed-off-by: Stine Akredalen <stine.akredalen@nordicsemi.no>

KKopyscinski committed:
tests: bluetooth: mesh: add Proxy Solicitation to shell

This adds commands to advertise Solicitation PDUs, configure
On-Demand Private Proxy and manage Solicitation PDU RPL.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

akredalen committed:
tests: Bluetooth: mesh: added LCD bsim test

Testing the basic behaviour of the Large Composition Data model.

Signed-off-by: Stine Akredalen <stine.akredalen@nordicsemi.no>

alxelax committed:
Bluetooth: Mesh: add mesh 1.1 global config option

All mesh 1.1 features are gathered behind a KConfig options.
Default is disabled.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

alxelax committed:
tests: Bluetooth: Mesh: run bsim mesh on both v1.0.1 and v1.1

Add ability to run bsim mesh for both ble mesh versions
v1.0.1 and v1.1

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: DFU and BLOB models in shell

Adds conditional support for the BLOB and DFU models in the mesh shell.
Additionally creates a dummy BLOB stream for testing.

Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Co-authored-by: Anders Storrø <anders.storro@nordicsemi.no>
Co-authored-by: Ludvig Samuelsen Jordet <ludvig.jordet@nordicsemi.no>
Co-authored-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>
Co-authored-by: Mia Koen <mia.koen@nordicsemi.no>
Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

Commit:
Bluetooth: Mesh: Transport SAR Configuration

Add support for:
- SAR Configuration Client and Server models
- Transport SAR configuration states
- Using SAR Transmitter/Receiver states in segmentation
and reassembly process.

Bsim tests fixes:
- Fix failing replay attack test. The replay attack test doesn't
consider retransmission attempts at the transport layer. When
retransmission happens, SeqNums get increased and the test logic
doesn't work anymore. The simplest fix would be to disable
retransmissions at the transport layer.
- Add device synchronization API to synchronize transport va test.
Device configuration may take different time on transmitter and
receiver. Add synchronisation barrier between devices.
- Fix msg_frnd test. Timing in transport sar behavior has been
changed, which affected test_friend_msg test. Now acknowledgments
are sent much faster and this needs to be considered in the test.
- Fix unicast_low_lat test. Increase number of retransmission
attempts when long segmented message is sent.
- Reduce timeout in kr_old_key test.
- Relax SAR RX state configuration in DFU and BLOB tests.

Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Co-authored-by: Anders Storrø <anders.storro@nordicsemi.no>
Co-authored-by: Ingar Kulbrandstad <ingar.kulbrandstad@nordicsemi.no>
Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
tests: bluetooth: mesh: Test confirm step in DFU with different effects

This commit implements the following tests:
- Test that DFU finishes successfully with all different effects;
- Test that DFU fails at confirm step with all different effects;

Co-authored-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
tests: bluetooth: mesh: Check DFU Client persistency

Check that DFU client continues DFU procedure as long as there is at
least one active target.

Co-authored-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
tests: Bluetooth: Mesh: Add distributor self update tests

Add a test where distributor is a single target and performs self
update. Add a test where distributor performs self update with multiple
targets.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

KKopyscinski committed:
tests: Bluetooth: Mesh: Test SAR with slowest possible timings

Add test that sends longest possible PDU with slowest possible timings
and default retransmission values.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

akredalen committed:
tests: Bluetooth: Mesh: add LCD metadata test. Refactor existing tests.

* Test that the Large Composition Data server can send a max access
payload metadata status message.
* Test that the Large Composition Data client can request large chunks
of metadata from the server, and correctly merge samples.
* Refactoring of existing LCD tests.

Signed-off-by: Stine Akredalen <stine.akredalen@nordicsemi.no>

omkar3141 committed:
Bluetooth: Mesh: Update mesh shell's config

This updates some of the defaults in the mesh shell's config to allow
various Mesh 1.1 models to be used in the shell.

EPA provisioning and DFU requires much larger multisegment messages,
therefore support for maximum sized message is enabled by default and
stack sizes are increased working values.

Additionally, log buffer is increased and thread sleep time is reduced
to avoid dropping of useful logs.

Signed-off-by: Omkar Kulkarni <omkar.kulkarni@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

LingaoM commented:
Done this feature public now?

PavelVPV commented:
> Done this feature public now?

The specs are publicly available here: https://www.bluetooth.com/specifications/specs/?types=specification&keyword=mesh&filter=

alxelax commented on tests/bluetooth/mesh_shell/testcase.yaml:
```diff
@@ -3,9 +3,7 @@ common:
-    platform_allow: qemu_x86 nrf51dk_nrf51422
-    integration_platforms:
-      - qemu_x86
+    platform_allow: qemu_cortex_m3 qemu_x86 nrf51dk_nrf51422 nrf52dk_nrf52832
```
Seems this unnecessary change sneaked somehow. The initial upstream file is more correct. I guess changes in this file should be reverted.

Thalley commented:
@PavelVPV Are those specification public? Zephyr shouldn't contain code for non-public specification IMO

PavelVPV commented:
> https://www.bluetooth.com/specifications/specs/?types=specification&keyword=mesh&filter=

Yes, they are public. Anyone can download them here: https://www.bluetooth.com/specifications/specs/?types=specification&keyword=mesh&filter=
Though the specs are in draft state yet.

LingaoM commented:
> > @PavelVPV Are those specification public? Zephyr shouldn't contain code for non-public specification IMO
> 
> 
> 
> Yes, they are public. Anyone can download them here: https://www.bluetooth.com/specifications/specs/?types=specification&keyword=mesh&filter=
> 
> Though the specs are in draft state yet.

It's best to put this information in the initial description so it's self-explanatory.

PavelVPV commented:
> It's best to put this information in the initial description so it's self-explanatory.

Links are added

jhedberg commented:
Lines added/removed: `+32,564 −809`. That's one of the biggest single pull requests I've seen in awhile, and will make the review process quite long and cumbersome. Any chance you could split this up into smaller, easier to review, parts, e.g. one PR per new feature?

PavelVPV commented:
> Lines added/removed: `+32,564 −809`. That's one of the biggest single pull requests I've seen in awhile, and will make the review process quite long and cumbersome. Any chance you could split this up into smaller, easier to review, parts, e.g. one PR per new feature?

We'd like to not split this PR into multiple ones. But the changes are split in commits per each feature. Shell commands for each feature are in separate commits. And BabbleSim tests are split in separate commits.

Balaklaka commented:
The features that has been added should be added to release notes.

jhedberg reviewed (commented):
Some minor things spotted after a *very* quick read-through. Not sure they're severe enough to block the PR for, but definitely things that should be eventually fixed.

Balaklaka commented:
Html doc can be found there: https://builds.zephyrproject.io/zephyr/pr/55393/docs/connectivity/bluetooth/api/mesh/models.html

PavelVPV committed:
docs: release: 3.4: Add exp support for new Bluetooth mesh draft specs

Add a note about added experimental support for new Bluetooth mesh
specifications.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
